src/rosella.rs

```diff
@@ -51,14 +57,71 @@ impl Rosella {
     }
 }
 
-impl Drop for Rosella {
+pub struct InstanceContext {
+    entry: ash::Entry,
+    instance: ash::Instance,
+}
+
+impl InstanceContext {
+    fn new(entry: ash::Entry, instance: ash::Instance) -> Self {
+        Self{ entry, instance }
+    }
+
+    pub fn get_entry(&self) -> &ash::Entry {
+        &self.entry
+    }
+
+    pub fn vk(&self) -> &ash::Instance {
+        &self.instance
+    }
+}
+
+impl Drop for InstanceContext {
     fn drop(&mut self) {
         unsafe {
-            self.device.destroy_device(ALLOCATION_CALLBACKS);
-            self.surface
-                .ash_surface
-                .destroy_surface(self.surface.khr_surface, ALLOCATION_CALLBACKS);
             self.instance.destroy_instance(ALLOCATION_CALLBACKS);
         }
     }
 }
+
+pub struct DeviceContext {
+    #[allow(unused)]
+    instance: Arc<InstanceContext>,
+    device: ash::Device,
+    synchronization_2: ash::extensions::khr::Synchronization2,
+    timeline_semaphore: ash::extensions::khr::TimelineSemaphore,
+}
+
+impl DeviceContext {
+    fn new(instance: Arc<InstanceContext>, device: ash::Device) -> Result<Self, &'static str> {
```

Comment:
Result? Only `Ok` is returned here

Reply:
I originally wanted to add additional verification to ensure all the extension functions have loaded properly.
Kept the Result since it seemed useful for potential future stuff but yeah it might be better to just remove it. Not sure.